src/OpenTelemetry.Api/Context/RuntimeContext.cs

```diff
@@ -27,10 +27,40 @@ public static class RuntimeContext
     {
         private static readonly ConcurrentDictionary<string, object> Slots = new();
 
+        private static Type contextSlotType = typeof(AsyncLocalRuntimeContextSlot<>);
+
         /// <summary>
         /// Gets or sets the actual context carrier implementation.
         /// </summary>
-        public static Type ContextSlotType { get; set; } = typeof(AsyncLocalRuntimeContextSlot<>);
+        public static Type ContextSlotType
+        {
+            get => contextSlotType;
+            set
+            {
+                Guard.ThrowIfNull(value, nameof(value));
+
+                if (value == typeof(AsyncLocalRuntimeContextSlot<>))
+                {
+                    contextSlotType = value;
+                }
+                else if (value == typeof(ThreadLocalRuntimeContextSlot<>))
+                {
+                    contextSlotType = value;
+                }
+#if NETFRAMEWORK
+                else if (value == typeof(RemotingRuntimeContextSlot<>))
+                {
+                    contextSlotType = value;
+                }
+#endif
+                else
+                {
+                    throw new NotSupportedException($"{value} is not a supported type.");
+                }
```

Comment:
Should it really throw Exception? As I remember OTel SDK/API should rather silently ignore non-working parameters than throwing exception.

Just to double check if Event/Log is not a better option.

Reply:
In this case, this isn't a "someone configured it wrong" or "something went wrong" situation. This is a "the developer used the API incorrectly" situation. So I think it is better to let the developer know early that what they are trying to do isn't possible. (i.e. throw an exception).

Reply:
https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/error-handling.md#basic-error-handling-principles

_The API or SDK MAY fail fast and cause the application to fail on initialization_